jdaviz/configs/mosviz/plugins/parsers.py

```diff
@@ -513,130 +517,190 @@ def _get_source_identifiers_by_hdu(hdus, filepaths=None, header_keys=['SOURCEID'
-                  "2D Spectra C": spec2d_c,
-                  "2D Spectra R": spec2d_r,
-                  "1D Spectra C": spec1d_c,
-                  "1D Spectra R": spec1d_r
-                 }
-
-    # Convert from pathlib Paths back to strings
-    for key in file_lists:
-        file_lists[key] = [str(x) for x in file_lists[key]]
-    _warn_if_not_found(app, file_lists)
-
-    # Parse relevant information from source catalog
-    cat_fields = ["id", "sky_centroid.ra", "sky_centroid.dec"]
-    source_ids = []
-    ras = []
-    decs = []
-    image_add = []
+        elif fp.suffix == '.ecsv':
+            if label_dict[catalog_key]:
+                raise ValueError("source directory has more "
+                                 "than one source catalog")
+            else:
+                label_dict[catalog_key] = fp
 
-    pupil_id_dict = {}
+        else:
+            continue
 
-    # Retrieve source information
-    for source_catalog_num in range(0, len(file_lists["Source Catalog"])):
-        cat_file = file_lists["Source Catalog"][source_catalog_num]
-        parsed_cat_fields = _fields_from_ecsv(cat_file, cat_fields, delimiter=" ")
-        pupil = [x
-                 for x in cat_file.split("/")[-1].split("_")
-                 if x[0] == "F" or x[0] == "f"][0]
+    return label_dict
 
-        pupil_id_dict[pupil] = {}
 
-        for row in parsed_cat_fields:
-            pupil_id_dict[pupil][int(row[0])] = (row[1], row[2])
+@data_parser_registry("mosviz-niriss-parser")
+def mos_niriss_parser(app, data_dir):
+    """
+    Attempts to parse all data for a NIRISS dataset in a single
+    directory, which should include:
 
-    # Read in direct image filters
-    image_dict = {}
-    filter_wcs = {}
+    - *_i2d.fits : Direct 2D images
+    - *_cat.ecsv : Source catalog
+    - *_cal.fits : 2D spectra in vertical (R) and horizontal (C)
+        orientations. C spectra are shown first in viewers by default
+    - *_x1d.fits : 1D spectra in vertical (R) and horizontal (C)
+        orientations. C spectra are shown first in viewers by default
 
-    # Set up a dictionary of datasets to add to glue
-    add_to_glue = {}
+    NOTE: For best performance, it's recommended that your directory
+    only contain one source catalog and its associated images/spectra.
 
-    print("Loading: Images")
+    Parameters
+    ----------
+    app : `~jdaviz.app.Application`
+        The application-level object used to reference the viewers.
+    data_dir : str
+        The path to a directory containing NIRISS data products.
+    """
+    path = Path(data_dir)
+    if not path.is_dir():
+        raise ValueError(f"{data_dir} is not a valid directory path")
+
+    # create dict for mapping expected file types to their DATAMODL identifiers
+    expected_labels = ['1D Spectra C', '1D Spectra R',
+                       '2D Spectra C', '2D Spectra R', 'Direct Image']
+    files_by_labels = {k: [] for k in expected_labels}
```

Comment:
Maybe `defaultdict` would simplify this setup, but depends on what you are trying to do, I guess.

Reply:
You actually taught me about `defaultdict` closer to the beginning of this pull request. I stopped using it because I didn't like the implication of `defaultdict(list)` for the source catalog key. The function only accepts one source catalog file, but a list can have more than one entry. I made that key's value a string instead.